File: bindings/python/examples/render.py
"""Render an image to ASCII art and print to the terminal.

Usage:
    python -m examples.render <image> <font> [options]

Examples:
    python -m examples.render ../../images/raccoon.jpg ../../fonts/GeistMono-Regular.ttf
    python -m examples.render ../../images/raccoon.jpg ../../fonts/GeistMono-Regular.ttf --color
    python -m examples.render ../../images/raccoon.jpg ../../fonts/GeistMono-Regular.ttf --color --cell-width 8 --cell-height 16
"""

import argparse
import sys

import glif


def main():
    parser = argparse.ArgumentParser(description="Render an image to ASCII art")
    parser.add_argument("image", help="Path to input image (PNG, JPEG, etc.)")
    parser.add_argument("font", help="Path to monospace TTF font")
    parser.add_argument("--color", action="store_true", help="ANSI truecolor output")
    parser.add_argument("--cell-width", type=int, default=10, help="Cell width in pixels (default: 10)")
    parser.add_argument("--cell-height", type=int, default=20, help="Cell height in pixels (default: 20)")
    parser.add_argument("--dir-crunch", type=float, default=1.25, help="Directional contrast exponent (default: 1.25)")
    parser.add_argument("--global-crunch", type=float, default=1.5, help="Global contrast exponent (default: 1.5)")
    args = parser.parse_args()

    result = glif.render(
        args.image, args.font,
        cell_width=args.cell_width,
        cell_height=args.cell_height,
        dir_crunch=args.dir_crunch,
        global_crunch=args.global_crunch,
    )

    if args.color:
        sys.stdout.write(result.to_ansi())
    else:
        print(result.to_plain_text())

    print(f"\n{result.cols}x{result.rows} grid", file=sys.stderr)


if __name__ == "__main__":
    main()

File: bindings/node/examples/render.mjs
/**
 * Render an image to ASCII art and print to the terminal.
 *
 * Usage:
 *   node render.mjs <image> <font> [--color] [--cell-width N] [--cell-height N]
 *
 * Examples:
 *   npm start                          # renders raccoon with default font
 *   node render.mjs ../../images/raccoon.jpg ../../fonts/GeistMono-Regular.ttf
 *   node render.mjs ../../images/raccoon.jpg ../../fonts/GeistMono-Regular.ttf --color
 *   node render.mjs ../../images/raccoon.jpg ../../fonts/GeistMono-Regular.ttf --color --cell-width 8 --cell-height 16
 */

import { createRequire } from 'node:module';
import { parseArgs } from 'node:util';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const require = createRequire(import.meta.url);
const glif = require('@artalis/glif');

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const { values, positionals } = parseArgs({
    allowPositionals: true,
    options: {
        color:         { type: 'boolean', default: false },
        'cell-width':  { type: 'string', default: '10' },
        'cell-height': { type: 'string', default: '20' },
        'dir-crunch':  { type: 'string', default: '1.25' },
        'global-crunch': { type: 'string', default: '1.5' },
    },
});

const imagePath = positionals[0] || path.join(__dirname, '..', '..', '..', 'images', 'raccoon.jpg');
const fontPath = positionals[1] || path.join(__dirname, '..', '..', '..', 'fonts', 'GeistMono-Regular.ttf');

const result = glif.render(imagePath, fontPath, {
    cellWidth: parseInt(values['cell-width']),
    cellHeight: parseInt(values['cell-height']),
    dirCrunch: parseFloat(values['dir-crunch']),
    globalCrunch: parseFloat(values['global-crunch']),
});

if (values.color) {
    process.stdout.write(result.toAnsi());
} else {
    console.log(result.toPlainText());
}

process.stderr.write(`\n${result.cols}x${result.rows} grid\n`);
